org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceVariable.java

```diff
@@ -209,14 +209,20 @@ public String toString() {
 		return debugName();
 	}
 
+	private int hashCode;
 	@Override
 	public int hashCode() {
-		int code = this.typeParameter.hashCode() + 17 * this.rank;
-		if (this.site != null) {
-			code = 31 * code + this.site.sourceStart();
-			code = 31 * code + this.site.sourceEnd();
+		if (this.hashCode == 0 || (this.typeParameter instanceof WildcardBinding
```

Comment:
I don't understand the part with `WildcardBinding`.
If the code is OK, probably this deserves an extra comment in the code.

But it looks like we assume that calling `this.typeParameter.hashCode()` may result in two different values for `WildcardBinding`? This would be not OK.

So it looks like we **expect** that `hashCode()` contract is broken for `WildcardBinding` (lazy loading or something like this) and so we only try to cache values if that is not the case, right?

That should be documented here and in `WildcardBinding` and if it can't be fixed in this PR, ideally we should create a dedicated bug and discuss there why `WildcardBinding.hashCode()` is broken and how that can be fixed.